### PR TITLE
WIP: More collection tools

### PR DESCRIPTION
**Pull request by vakabus**

### CODE REVIEW

**Comment by vakabus** [comment]
@rqu The code is not well documented. But there is nothing complicated there. What do you think about it?

File: blood/src/main/java/cz/cuni/mff/d3s/blood/node_origin_tracker/DependencyMatrixCollector.java [hunk 1]
```diff
@@ -56,6 +57,11 @@
         phaseOrder.add(NodeTracker.DeletedPhaseDummy.class);
     }
 
+    public DependencyMatrixCollector() {
+        // register hook for dumping data
+        Report.getInstance().registerDump(new ManualDump("depmat", this::dump));
```
Comment (rqu) [inline blood/src/main/java/cz/cuni/mff/d3s/blood/node_origin_tracker/DependencyMatrixCollector.java]: I think constructors should not have side effects. Just a general rule. I don't know, how it applies on singletons.

(Although onNodeClassInit is not a good place for this code, either.)

File: blood/src/main/java/cz/cuni/mff/d3s/blood/node_type_list/NodeTypeList.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.stream.Collectors;
+
+public class NodeTypeList {
+    private static NodeTypeList instance = null;
+    private Set<Class> nodeClasses = Collections.newSetFromMap(new ConcurrentHashMap<>());
+    private ReadWriteLock rwLock = new ReentrantReadWriteLock(true);
+
+    public NodeTypeList() {
+        // register dump
```
Comment (rqu) [inline blood/src/main/java/cz/cuni/mff/d3s/blood/node_type_list/NodeTypeList.java]: `<nitpick>`Unhelpful comment`</nitpick>`

File: blood/src/main/java/cz/cuni/mff/d3s/blood/node_origin_tracker/DependencyMatrixCollector.java [hunk 3]
```diff
@@ -56,6 +57,11 @@
         phaseOrder.add(NodeTracker.DeletedPhaseDummy.class);
     }
 
+    public DependencyMatrixCollector() {
```
Comment (rqu) [inline blood/src/main/java/cz/cuni/mff/d3s/blood/node_origin_tracker/DependencyMatrixCollector.java]: Should the constructor not be private?

Comment (rqu) [reply blood/src/main/java/cz/cuni/mff/d3s/blood/node_origin_tracker/DependencyMatrixCollector.java]: Reading further, I see that this is not the only case. I will not comment.

File: blood/src/main/java/cz/cuni/mff/d3s/blood/node_origin_tracker/DependencyMatrixCollector.java (hunk 1 above)
Comment (rqu) [reply blood/src/main/java/cz/cuni/mff/d3s/blood/node_origin_tracker/DependencyMatrixCollector.java]: Again, this is not the only case.

File: blood/src/main/java/cz/cuni/mff/d3s/blood/node_type_tracker/NodeTypeTrackerCollector.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+import cz.cuni.mff.d3s.blood.report.Report;
+import cz.cuni.mff.d3s.blood.report.dump.ManualDump;
+import org.graalvm.compiler.graph.Node;
+import org.graalvm.compiler.nodes.StructuredGraph;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class NodeTypeTrackerCollector {
```
Comment (rqu) [inline blood/src/main/java/cz/cuni/mff/d3s/blood/node_type_tracker/NodeTypeTrackerCollector.java:17]: Copy-pasted from `DependencyMatrixCollector`. Maybe extract the matrix code to a `Matrix<ValueType>` class?

File: blood/src/main/java/cz/cuni/mff/d3s/blood/node_type_tracker/NodeTypeTrackerCollector.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    public static NodeTypeTrackerCollector getInstance() {
+        if (instance == null) {
+            instance = new NodeTypeTrackerCollector();
+        }
+        return instance;
+    }
+
+    public final void updateNodeTypes(StructuredGraph graph, ConcurrentHashMap<Class, NodeTrackerValue> phaseTable) {
```
Comment (rqu) [inline blood/src/main/java/cz/cuni/mff/d3s/blood/node_type_tracker/NodeTypeTrackerCollector.java]: A bit misleading name, but self-explanatory code. What about `updateRow`?

File: blood/src/main/java/cz/cuni/mff/d3s/blood/node_type_tracker/NodeTypeTrackerCollector.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        HashMap<Class, Long> nodeCount = new HashMap<>();
+        for (Node n : graph.getNodes()) {
+            Long count = nodeCount.getOrDefault(n.getClass(), 0l);
```
Comment (rqu) [inline blood/src/main/java/cz/cuni/mff/d3s/blood/node_type_tracker/NodeTypeTrackerCollector.java:50]: Typo

File: blood/src/main/java/cz/cuni/mff/d3s/blood/report/Report.java [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    public void registerDump(DumpRegistration registration) {
+        registrations.add(registration);
+    }
+
+    private void onShutdown() {
+        var startTime = Instant.now();
+
+        new File(DumpHelpers.DUMPS_DIR_NAME).mkdir();
```
Comment (rqu) [inline blood/src/main/java/cz/cuni/mff/d3s/blood/report/Report.java]: Already in `DumpHelpers::getDumpFileWriter`. I suggest removing it from there and leaving it here.

File: blood/src/main/java/cz/cuni/mff/d3s/blood/report/Report.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+
+        String timestamp = DumpHelpers.getDateString();
+        File reportDir = new File(DumpHelpers.DUMPS_DIR_NAME + "/report-" + timestamp);
```
Comment (rqu) [inline blood/src/main/java/cz/cuni/mff/d3s/blood/report/Report.java]: I find the approach used in `DumpHelpers::getDumpFileWriter` (i.e. `new File(File parent, String child)`) more elegant than concatenating with a string containing directory separator.

File: blood/src/main/java/cz/cuni/mff/d3s/blood/report/Report.java [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+        for (var dr : registrations) {
+            LOGGER.log(System.Logger.Level.INFO, "Dumping {0}", dr.getName());
+
+            var outfile = new File(reportDir, dr.getName());
+            var data = dr.dumpData();
+            try (var fos = new FileOutputStream(outfile)) {
+                fos.write(data);
+            } catch (IOException e) {
+                System.err.println("Failed to dump report " + dr.getName());
+                e.printStackTrace();
```
Comment (rqu) [inline blood/src/main/java/cz/cuni/mff/d3s/blood/report/Report.java]: I would use the logger for consistency.
```java
LOGGER.log(Level.???, "Failed to dump report " + dr.getName(), e);
```

File: blood/src/main/java/cz/cuni/mff/d3s/blood/report/dump/ManualDump.java [hunk 10]
```diff
@@ -0,0 +1,23 @@
+package cz.cuni.mff.d3s.blood.report.dump;
+
+import java.util.function.Supplier;
+
+public final class ManualDump implements DumpRegistration {
+    final String name;
+    final Supplier<byte[]> dumpProducer;
```
Comment (rqu) [inline blood/src/main/java/cz/cuni/mff/d3s/blood/report/dump/ManualDump.java]: Why are you using `byte[]` when most (currently all) usages generate text? What about (at least) making `String` (or `CharSequence`) an option through another `DumpRegistration` implementation?

File: blood/src/main/java/cz/cuni/mff/d3s/blood/report/Report.java [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.List;
+
+public final class Report {
```
Comment (rqu) [inline blood/src/main/java/cz/cuni/mff/d3s/blood/report/Report.java:12]: Nice :) 2/3 of #4 are done.

**Review by rqu — COMMENTED** [review]
One comment that did not fit anywhere else:
In package name, `node_origin_tracker` looks better than `dependencyMatrix` (although less Java :) ), but the choice of words is unhappy.
First, it diverges from all the names that are related to it, like `class DependencyMatrixCollector` and `"depmat"`.
Second, it collides with the meaning of the name `interface NodeTracker`.

Don't take this as a negative review. Consider all lines without a comment acclaimed :)

File: blood/src/main/java/cz/cuni/mff/d3s/blood/node_type_tracker/NodeTypeTrackerCollector.java (hunk 6 above)
Comment (vakabus) [reply blood/src/main/java/cz/cuni/mff/d3s/blood/node_type_tracker/NodeTypeTrackerCollector.java:50]: I don't see it. If you mean the `0l` constant, that's type specification placed there on purpose...

File: blood/src/main/java/cz/cuni/mff/d3s/blood/node_origin_tracker/DependencyMatrixCollector.java (hunk 1 above)
Comment (vakabus) [reply blood/src/main/java/cz/cuni/mff/d3s/blood/node_origin_tracker/DependencyMatrixCollector.java]: Kind of agree, but this is the best place I could come up with. :man_shrugging:  I also think, that it makes sense semantically. Creating multiple collectors should automatically result in multiple dumps happening. (Which will fail due to Report implementation, but that could be fixed when necessary.)

So - won't fix. Don't know how.

File: blood/src/main/java/cz/cuni/mff/d3s/blood/node_origin_tracker/DependencyMatrixCollector.java (hunk 3 above)
Comment (vakabus) [reply blood/src/main/java/cz/cuni/mff/d3s/blood/node_origin_tracker/DependencyMatrixCollector.java]: Yup. Thanks, will fix.

File: blood/src/main/java/cz/cuni/mff/d3s/blood/node_type_tracker/NodeTypeTrackerCollector.java (hunk 4 above)
Comment (vakabus) [reply blood/src/main/java/cz/cuni/mff/d3s/blood/node_type_tracker/NodeTypeTrackerCollector.java:17]: This won't be the case soon. `DependencyMatrixCollector`'s matrix will change quite a bit in a short amount of time due to planned changes. Probably won't fix.

File: blood/src/main/java/cz/cuni/mff/d3s/blood/node_type_tracker/NodeTypeTrackerCollector.java (hunk 5 above)
Comment (vakabus) [reply blood/src/main/java/cz/cuni/mff/d3s/blood/node_type_tracker/NodeTypeTrackerCollector.java]: Good point.

File: blood/src/main/java/cz/cuni/mff/d3s/blood/report/Report.java (hunk 7 above)
Comment (vakabus) [reply blood/src/main/java/cz/cuni/mff/d3s/blood/report/Report.java]: Ok. Will be fixed.

File: blood/src/main/java/cz/cuni/mff/d3s/blood/report/Report.java (hunk 8 above)
Comment (vakabus) [reply blood/src/main/java/cz/cuni/mff/d3s/blood/report/Report.java]: You are right. Might be even better to look at new Path API...

**Comment by vakabus** [comment]
I got rid of that package name because of semantical conflict with NodeTypeTracker. It too creates a matrix and it too describes some sort of dependency between node types and phases. I would consider renaming it all to something better. Not just the package.

Other issues were addressed in my last commit.

File: blood/src/main/java/cz/cuni/mff/d3s/blood/node_type_tracker/NodeTypeTrackerCollector.java (hunk 6 above)
Comment (rqu) [reply blood/src/main/java/cz/cuni/mff/d3s/blood/node_type_tracker/NodeTypeTrackerCollector.java:50]: Sorry. In the font GitHub UI uses, it seemed to me like "zero one". I thought you also pressed 1 while aiming at 0.

**Comment by vakabus** [comment]
Phase order won'ŧ work with such simple implementation due to inlining . I propose removing that feature and merging this PR into master. Because other than that, everything works. Then, I would work on phase order tracking in a different PR.

**Comment by vakabus** [comment]
Will be merged with #12

**Pull request closed without merging**